### PR TITLE
Goal refactor to allow for local image files

### PR DESCRIPTION
**Pull request by ljang0**
@ThibaultLSDC can you check this?

### CODE REVIEW

File: browsergym/core/src/browsergym/core/env.py [hunk 1]
```diff
@@ -277,9 +301,23 @@ def override_property(task, env, property):
...
+                    if image_url.startswith("http"):
+                        image = Image.open(requests.get(image_url, stream=True).raw)
+                        image_base64 = pil_to_b64(image)
+                    elif image_url.startswith("data:image/png;base64,"):
+                        image_base64 = image_url
+                        image = b64_to_pil(image_base64)
+                    else:
+                        raise ValueError(f"Unexpected image_url: {image_url}")
+                    # save the image to a temporary (but persistent) PNG file
+                    if not os.path.exists("tmp"):
```
Comment (gasse) [inline browsergym/core/src/browsergym/core/env.py:315]: This won't work when running agents in parallel, there will be name collisions.

**Review by gasse — CHANGES_REQUESTED** [review]

File: browsergym/core/src/browsergym/core/env.py (hunk 1 above)
Comment (ljang0) [reply browsergym/core/src/browsergym/core/env.py:315]: what do you think the workaround is? some benchmarks have the same template id across tasks as well, so using unique ids might cause name collisions as well

File: browsergym/core/src/browsergym/core/env.py [hunk 2]
```diff
@@ -508,26 +518,16 @@ def _get_obs(self):
...
+            logger.warning(f"Empty goal, trying to extract goal from chat as a fallback.")
+            self.goal = self._try_and_extract_goal_from_chat()
 
         # obs is generic to all tasks
         obs = {
             "chat_messages": copy.deepcopy(self.chat.messages),
-            "goal": goal_msg,  # TODO: redundant with chat messages, to be removed?
-            "goal_image_urls": goal_image_urls,  # TODO: redundant with chat messages, to be removed?
+            "goal": self._goal_to_text(self.goal),  # legacy goal, deprecated
+            "goal_object": self.goal,  # new goal format, liust of messages openai style
```
Comment (gasse) [inline browsergym/core/src/browsergym/core/env.py:530]:
```suggestion
            "goal_object": self.goal,  # new goal format, list of messages openai style
```

File: browsergym/experiments/src/browsergym/experiments/loop.py [hunk 3]
```diff
@@ -414,7 +415,18 @@ def make_stats(self):
...
+            # save the goal object only once (goal should never change once setup)
+            goal_object_file = exp_dir / "goal_object.pkl.gz"
+            if not os.path.exists(goal_object_file):
+                with gzip.open(goal_object_file, "wb") as f:
+                    pickle.dump(self.obs["goal_object"], f)
+            # set goal_object to a special placeholder value, which indicates it should be loaded from a separate file
+            self.obs["goal_object"] = None
+
         with gzip.open(exp_dir / f"step_{self.step}.pkl.gz", "wb") as f:
+            # TODO should we pop the screenshots too before this save space ?
```
Comment (gasse) [inline browsergym/experiments/src/browsergym/experiments/loop.py:429]:
```suggestion
            # TODO should we pop the screenshots too before this to save space ?
```

File: browsergym/visualwebarena/src/browsergym/visualwebarena/task.py [hunk 4]
```diff
@@ -129,26 +131,60 @@ def setup(self, page: playwright.sync_api.Page) -> tuple[str, dict]:
...
+        # textual goal first
+        goal = [{"type": "text", "text": goal_text}]
+        # then goal images
+        for i, (image_url, image_path) in enumerate(zip(image_urls, image_paths)):
+            goal.extend(
+                [
+                    # image description (id and filepath)
+                    {
+                        "type": "text",
+                        "text": f"Input image {i}/{len(image_urls)} below (local path: {repr(image_path)})",
```
Comment (gasse) [inline browsergym/visualwebarena/src/browsergym/visualwebarena/task.py:176]:
```suggestion
                        "text": f"Input image {i+1}/{len(image_urls)} below (local path: {repr(image_path)})",
```

File: browsergym/core/src/browsergym/core/env.py [hunk 5]
```diff
@@ -508,26 +518,16 @@ def _get_obs(self):
...
+        if not self.goal:
+            logger.warning(f"Empty goal, trying to extract goal from chat as a fallback.")
+            self.goal = self._try_and_extract_goal_from_chat()
 
         # obs is generic to all tasks
         obs = {
             "chat_messages": copy.deepcopy(self.chat.messages),
-            "goal": goal_msg,  # TODO: redundant with chat messages, to be removed?
-            "goal_image_urls": goal_image_urls,  # TODO: redundant with chat messages, to be removed?
+            "goal": self._goal_to_text(self.goal),  # legacy goal, deprecated
```
Comment (recursix) [inline browsergym/core/src/browsergym/core/env.py:529]: minor: make this as a non class function. So it's easier to know inputs and
outputs and also easier to unit test.

File: browsergym/experiments/src/browsergym/experiments/loop.py [hunk 6]
```diff
@@ -414,7 +415,18 @@ def make_stats(self):
 
     def save_step_info(self, exp_dir, save_json=False, save_jpg=True):
 
+        # save goal object (which might contain images) to a separate file to save space
+        if self.obs is not None and self.obs.get("goal_object", False):
+            # save the goal object only once (goal should never change once setup)
+            goal_object_file = exp_dir / "goal_object.pkl.gz"
+            if not os.path.exists(goal_object_file):
```
Comment (recursix) [inline browsergym/experiments/src/browsergym/experiments/loop.py:422]: pathlib please :)

File: browsergym/visualwebarena/src/browsergym/visualwebarena/task.py [hunk 7]
```diff
@@ -129,26 +131,60 @@ def setup(self, page: playwright.sync_api.Page) -> tuple[str, dict]:
...
-            goal[
-                "message"
-            ] += """\
+            goal_text += """\
 
 If you believe the task is impossible to complete, provide the answer "N/A".
 """
 
+        # recover goal image urls
+        image_urls = self.config.get("image", [])
```
Comment (recursix) [inline browsergym/visualwebarena/src/browsergym/visualwebarena/task.py:145]: this my not work with remote instances of VWA

Comment (recursix) [reply browsergym/visualwebarena/src/browsergym/visualwebarena/task.py:145]: implementing a sandbox directory could be a solution

**Review by recursix — COMMENTED** [review]
Great job, But I'm really disappointed in the number of unit tests :p you will regret it.

**Review by recursix — DISMISSED** [review]

File: browsergym/experiments/src/browsergym/experiments/loop.py (hunk 6 above)
Comment (gasse) [reply browsergym/experiments/src/browsergym/experiments/loop.py]: done

File: browsergym/core/src/browsergym/core/env.py (hunk 5 above)
Comment (gasse) [reply browsergym/core/src/browsergym/core/env.py]: done

File: browsergym/experiments/src/browsergym/experiments/loop.py [hunk 8]
```diff
@@ -412,7 +413,18 @@ def make_stats(self):
 
     def save_step_info(self, exp_dir, save_json=False, save_jpg=True):
 
+        # save goal data (which might contain images) to a separate file to save space
+        if self.obs is not None and self.obs.get("goal_data", None) is not None:
+            # save the goal data only once (goal should never change)
```
[the remark this answers is not in the document]
Comment (gasse) [reply browsergym/experiments/src/browsergym/experiments/loop.py]: yes

File: demo_agent/basic_agent.py [hunk 9]
```diff
@@ -0,0 +1,330 @@
...
+                    user_msgs.append(
+                        {
+                            "type": "text",
+                            "text": f"""\
+- [{msg['role']}] {msg['message']}
+""",
+                        }
+                    )
+                elif msg["role"] == "user_image":
+                    system_msgs.append({"type": "image_url", "image_url": msg["message"]})
```
Comment (ThibaultLSDC) [inline demo_agent/basic_agent.py:124]:
```suggestion
                    user_msgs.append({"type": "image_url", "image_url": msg["message"]})
```
?

**Review by gasse — APPROVED** [review]

**Pull request merged**